execution-engine/src/wasi/wasmi.rs

```diff
@@ -0,0 +1,1203 @@
+                Self::POINTER,
+                Self::SIZE_T,
+                Self::FD,
+                Self::POINTER,
+                Self::SIZE_T,
+            ],
+            WASIAPIName::PATH_UNLINK_FILE => vec![Self::FD, Self::POINTER, Self::SIZE_T],
+            WASIAPIName::POLL_ONEOFF => vec![
+                Self::CONST_POINTER,
+                Self::POINTER,
+                Self::SIZE,
+                Self::POINTER,
+            ],
+            WASIAPIName::PROC_EXIT => vec![Self::EXITCODE],
+            WASIAPIName::PROC_RAISE => vec![Self::SIGNAL],
+            WASIAPIName::SCHED_YIELD => vec![],
+            WASIAPIName::RANDOM_GET => vec![Self::POINTER, Self::SIZE],
+            WASIAPIName::SOCK_RECV => vec![
+                Self::FD,
+                Self::IOVEC_ARRAY_BASE,
+                Self::IOVEC_ARRAY_LENGTH,
+                Self::RIFLAGS,
+                Self::POINTER,
+                Self::POINTER,
+            ],
+            WASIAPIName::SOCK_SEND => vec![
+                Self::FD,
+                Self::CIOVEC_ARRAY_BASE,
+                Self::CIOVEC_ARRAY_LENGTH,
+                Self::SIFLAGS,
+                Self::POINTER,
+            ],
+            WASIAPIName::SOCK_SHUTDOWN => vec![Self::FD, Self::SDFLAGS],
+        }
+    }
+
+    pub(crate) fn check_args_number(
+        args: &RuntimeArgs,
+        index: WASIAPIName,
+    ) -> Result<(), FatalEngineError> {
+        if args.len() == Self::get_params(index).len() {
+            Ok(())
+        } else {
+            Err(index.into())
+        }
+    }
+}
+
+/// Checks the signature of the module's entry point, `signature`, against the
+/// templates described above for the `EntrySignature` enum type, and returns
+/// an instance of that type as appropriate.
+fn check_main_signature(signature: &Signature) -> EntrySignature {
+    let params = signature.params();
+    let return_type = signature.return_type();
+
+    if params == [] && return_type == None {
+        EntrySignature::NoParameters
+    } else if params == [ValueType::I32, ValueType::I32] && return_type == None {
+        EntrySignature::ArgvAndArgc
+    } else {
+        EntrySignature::NoEntryFound
+    }
+}
+
+/// Finds the entry point of the WASM module, `module`, and extracts its
+/// signature template.  If no entry is found returns
+/// `EntrySignature::NoEntryFound`.
+fn check_main(module: &ModuleInstance) -> EntrySignature {
+    match module.export_by_name(WASIWrapper::ENTRY_POINT_NAME) {
+        Some(ExternVal::Func(funcref)) => check_main_signature(&funcref.signature()),
+        _otherwise => EntrySignature::NoEntryFound,
+    }
+}
+
+/// Finds the linear memory of the WASM module, `module`, and returns it,
+/// otherwise creating a fatal host error that will kill the Veracruz instance.
+fn get_module_memory(module: &ModuleRef) -> Result<MemoryRef, FatalEngineError> {
+    match module.export_by_name(WASIWrapper::LINEAR_MEMORY_NAME) {
```

Comment:
Neater written using `map` and `map_err`?  Also, another candidate for inlining.

Reply:
I do not think so, here the match is on `Some(ExternVal::Memory(memoryref))` rather than `Some(memoryref)`. Note that there could be  `Some(ExternVal::OtherThing(_))`.